code/game/objects/items/RCD.dm

```diff
@@ -670,81 +533,136 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
-		choices += list(
-		"Silo Link" = image(icon = 'icons/obj/mining.dmi', icon_state = "silo"),
-		)
-	if(upgrade & RCD_UPGRADE_FURNISHING)
-		choices += list(
-		"Furnishing" = image(icon = 'icons/hud/radial.dmi', icon_state = "chair")
-		)
-	switch(construction_mode)
-		if(RCD_AIRLOCK)
-			choices += list(
-			"Change Access" = image(icon = 'icons/hud/radial.dmi', icon_state = "access"),
-			"Change Airlock Type" = image(icon = 'icons/hud/radial.dmi', icon_state = "airlocktype")
-			)
-		if(RCD_WINDOWGRILLE)
-			choices += list(
-			"Change Window Glass" = image(icon = 'icons/hud/radial.dmi', icon_state = "windowtype"),
-			"Change Window Size" = image(icon = 'icons/hud/radial.dmi', icon_state = "windowsize")
-			)
-		if(RCD_FURNISHING)
-			choices += list(
-			"Change Furnishing Type" = image(icon = 'icons/hud/radial.dmi', icon_state = "chair")
-			)
-	var/choice = show_radial_menu(user, src, choices, custom_check = CALLBACK(src, PROC_REF(check_menu), user), require_near = TRUE, tooltips = TRUE)
-	if(!check_menu(user))
+
+/obj/item/construction/rcd/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "RapidConstructionDevice", name)
+		ui.open()
+
+/**
+ * if ui_always_active = TRUE display & update window even if nothing changed, required for construction_console else window wont show up
+ * else use parent method to decide the state for normal usage
+ */
+/obj/item/construction/rcd/ui_state(mob/user)
+	return ui_always_active ? GLOB.always_state : ..()
+
+/obj/item/construction/rcd/ui_static_data(mob/user)
+	return airlock_electronics.ui_static_data(user)
+
+/obj/item/construction/rcd/ui_data(mob/user)
+	var/list/data = ..(user)
+
+	data["space"] = TRUE
+	//main categories
+	data["selected_root"] = root_category
+	data["root_categories"] = list()
+	for(var/category in root_categories)
+		data["root_categories"] += category
+	data["root_categories"] += "AirLock Access"
+
+	//create the category list
+	data["selected_category"] = design_category
+	data["selected_design"] = design_title
+	data["categories"] = list()
+
+	for(var/list/sub_category as anything in root_categories[root_category])
+		//skip category if upgrades were not installed for these
+		if(sub_category == "Machines" && !(upgrade & RCD_UPGRADE_FRAMES))
+			continue
+		if(sub_category == "Furniture" && !(upgrade & RCD_UPGRADE_FURNISHING))
+			continue
+
+		var/list/designs = list() //initialize all designs under this category
+		var/list/target_category =  root_categories[root_category][sub_category]
+		for(var/i in 1 to target_category.len)
+			var/list/design = target_category[i]
+
+			//get/infer icon_state based on category
+			var/icon
+			if(sub_category == "Solid AirLocks")
+				icon = "[design["title"]]"
+			else if(sub_category == "Glass AirLocks")
+				icon = "[design["title"]]-Glass"
+			else
+				icon = design["icon"]
+			icon = replacetext(icon," ","-") //css class names cannot have spaces in them
```

Comment:
I do not understand why you don't want this to be `sanitize_css_class_name`??

Reply:
Done. I did use it but it didn't work, then i realized i had to do the same in the asset file